File: 백준/Gold/1029. 그림 교환/그림 교환.py
import sys
input = sys.stdin.readline

N = int(input())
price = [list(map(int, input().strip())) for _ in range(N)]

INF = 10
dp = [[INF] * N for _ in range(1 << N)]

# 시작 상태
dp[1][0] = 0  # 1번 아티스트, 가격 0

for mask in range(1 << N):
    for last in range(N):
        if dp[mask][last] == INF:
            continue

        cur_price = dp[mask][last]

        for nxt in range(N):
            if mask & (1 << nxt):
                continue
            if price[last][nxt] >= cur_price:
                new_mask = mask | (1 << nxt)
                dp[new_mask][nxt] = min(dp[new_mask][nxt],
                                         price[last][nxt])

ans = 0
for mask in range(1 << N):
    for last in range(N):
        if dp[mask][last] != INF:
            ans = max(ans, bin(mask).count('1'))

print(ans)

File: 백준/Gold/1029. 그림 교환/그림 교환.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        int[][] price = new int[N][N];
        for (int i = 0; i < N; i++) {
            String s = br.readLine();
            for (int j = 0; j < N; j++) {
                price[i][j] = s.charAt(j) - '0';
            }
        }

        final int INF = 10;
        int MAX_MASK = 1 << N;

        int[][] dp = new int[MAX_MASK][N];
        for (int i = 0; i < MAX_MASK; i++) {
            Arrays.fill(dp[i], INF);
        }

        // 시작 상태
        dp[1][0] = 0;

        for (int mask = 0; mask < MAX_MASK; mask++) {
            for (int last = 0; last < N; last++) {
                if (dp[mask][last] == INF) continue;

                int curPrice = dp[mask][last];

                for (int nxt = 0; nxt < N; nxt++) {
                    if ((mask & (1 << nxt)) != 0) continue;

                    if (price[last][nxt] >= curPrice) {
                        int newMask = mask | (1 << nxt);
                        dp[newMask][nxt] = Math.min(dp[newMask][nxt],
                                                    price[last][nxt]);
                    }
                }
            }
        }

        int ans = 0;
        for (int mask = 0; mask < MAX_MASK; mask++) {
            for (int last = 0; last < N; last++) {
                if (dp[mask][last] != INF) {
                    ans = Math.max(ans, Integer.bitCount(mask));
                }
            }
        }

        System.out.println(ans);
    }
}
